### PR TITLE
Fixes #23598 - move report metrics chart to c3

### PR DESCRIPTION
**Pull request by lizagilman**
Using pf-react donut chart instead of flot

image

### CODE REVIEW

File: app/views/config_reports/_metrics.html.erb [hunk 1]
```diff
@@ -4,9 +4,8 @@
   <div class="col-md-4">
     <div class="stats-well">
       <h4 class="ca" ><%= _('Report Metrics') %></h4>
-      <div style="margin-top:50px;padding-bottom: 40px;">
-        <%= flot_pie_chart("metrics" ,_("Report Metrics"), metrics, :class => "statistics-pie small")%>
-      </div>
+      <div id="metrics_chart" class='metrics-chart' style="margin-top:50px;padding-bottom: 40px;"></div>
```
Comment (amirfefer) [inline app/views/config_reports/_metrics.html.erb]: why not add the inline style to the `metrics-chart` class?

File: app/views/config_reports/_metrics.html.erb [hunk 2]
```diff
@@ -18,7 +17,7 @@
   <div class="col-md-4">
     <table class="<%= table_css_classes %>" style="height: 398px;">
       <tbody>
-        <% metrics.sort.each do |title, value|%>
+      <% metrics.sort.each do |title, value|%>
```
Comment (amirfefer) [inline app/views/config_reports/_metrics.html.erb]: indentation should be 2 space

File: app/views/config_reports/_metrics.html.erb [hunk 3]
```diff
@@ -27,12 +26,12 @@
...
             </td>
           </tr>
-        <% end %>
+      <% end %>
       </tbody>
       <tfoot>
-        <tr>
-          <th><%= _("Total") %></th><th><%= metric (@totaltime || @config_report.runtime) %></th>
-        </tr>
+      <tr>
```
Comment (amirfefer) [inline app/views/config_reports/_metrics.html.erb]: ditto

File: app/views/config_reports/_metrics.html.erb [hunk 4]
```diff
@@ -4,9 +4,8 @@
   <div class="col-md-4">
     <div class="stats-well">
       <h4 class="ca" ><%= _('Report Metrics') %></h4>
-      <div style="margin-top:50px;padding-bottom: 40px;">
-        <%= flot_pie_chart("metrics" ,_("Report Metrics"), metrics, :class => "statistics-pie small")%>
-      </div>
+      <div id="metrics_chart" class='metrics-chart' style="margin-top:50px;padding-bottom: 40px;"></div>
+      <%= mount_react_component('DonutChart', '#metrics_chart', metrics.to_a) %>
```
Comment (amirfefer) [inline app/views/config_reports/_metrics.html.erb]: data should be in json format (`to_json`)

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @lizagilman 
I left some inline comments.

We have a component for a chart box with this design:
chart-box
 although I know that this page uses chart bar which still using flot.

**Comment by lizagilman** [comment]
@amirfefer thanks for the quick review, updated.
should we move only one chart to a box? once @ripcurld is finished with the bar chart we can move both to chartbox

File: app/assets/stylesheets/charts.scss [hunk 5]
```diff
@@ -143,6 +143,13 @@
   font-size: 14px;
 }
 
+.metrics-chart {
+  width: 240px;
+  margin: 12px auto;
+  margin-top: 50px;
```
Comment (amirfefer) [inline app/assets/stylesheets/charts.scss]: duplicate `margin-top` with 12px and 50px values.
maybe use margin with 4 values?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @lizagilman , just one minor issue with the css.

Generally speaking, We do test the component behavior, 
but we don't test if a component actually is rendered in a page (end to end test).
do we want this kind of test?

**Comment by lizagilman** [comment]
@amirfefer thanks, css updated
I tried to test for the chart rendering both on integration test and controller test but it seems the container div is empty (the component is not mounted to `#metrics_chart`). Is this a normal behavior?

**Comment by amirfefer** [comment]
@lizagilman, in order to run js in integration test you should use js driver, tests with js end with `_js_test`, and inherit `IntegrationTestWithJavascript`.

**Comment by lizagilman** [comment]
@amirfefer ok, I added a test
Is this what you had in mind?

**Comment by ohadlevy** [comment]
@lizagilman mind adding an updated screenshot? thanks.

File: test/integration/config_report_js_test.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+require 'integration_test_helper'
+
+class ConfigReportJSTest < IntegrationTestWithJavascript
+  test "should display report metrics chart" do
+    @report = ConfigReport.import(read_json_fixture('reports/applied.json'))
```
Comment (amirfefer) [inline test/integration/config_report_js_test.rb]: I think `@` is redundant here

**Review by amirfefer — CHANGES_REQUESTED** [review]

**Comment by lizagilman** [comment]
before: 

metrics_flot

aftre:

metrics_c3

File: test/integration/config_report_js_test.rb (hunk 6 above)
Comment (lizagilman) [reply test/integration/config_report_js_test.rb]: fixed

**Comment by amirfefer** [comment]
The chart boxes are not at the same size: 
report_charts

I wonder if the right table with the metric values is needed,  or we can use donut chart  legend:
report_metrics

**Review by amirfefer — CHANGES_REQUESTED** [review]
Overall looks good :+1: 
just a minor adjustment is needed, see other comment for details

**Comment by lizagilman** [comment]
updated - 

image

**Review by amirfefer — DISMISSED** [review]
LGTM :+1:

**Comment by ohadlevy** [comment]
this looks like a solid change, the only thing I'm a bit uncomfortable with is the red (and to some degree green) color inside the chart, in your screenshots it looks like something requires attention, when i looked at it on some other sample data it didnt look as bad, but it still feels wrong to me, @rohoover any thoughts?
image
image
image

**Comment by amirfefer** [comment]
@ohadlevy: these colors are suggested by pf color-palette and mentioned [here](https://github.com/theforeman/foreman/pull/5137#issuecomment-356959608)
But I tend to agree when it's about two distinct colors, the red color has some negative impact.
We can change the color pattern order, and move the red color from the [second place](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/services/ChartService.consts.js#L21).

**Comment by Rohoover** [comment]
@amirfefer Is correct. There is some flexibility as the donut doesn't specify which colors and only gives guidance to use the recommended [PF Color palette](http://www.patternfly.org/styles/color-palette/). We can technically choose from any of those colors.

**Comment by ohadlevy** [comment]
so in this case would you recommend changing the colors or its just me?

**Comment by Rohoover** [comment]
@ohadlevy 

It's you. ;)

In all seriousness, I am not married to any particular color choice at this point. I think PF does use the red, orange, green colors as both color and status in the donuts, which can be a bit confusing.

**Comment by ares** [comment]
I'd also prefer to avoid red if that's easy to do. But I can live with it :-)

**Comment by lizagilman** [comment]
How about now? 

1. red is last

image

image

2.  or we can replace it with a different color -

image


for now, pr updated with option 1 (red at last place)

@ohadlevy @ares do you prefer to avoid red completely?

**Comment by ares** [comment]
I prefer option 2, but not strong opinion. If red helps color-blind people, I'm fine with keeping it.

**Comment by lizagilman** [comment]
@tstrachota any preferences? or can we merge?

**Review by amirfefer — CHANGES_REQUESTED** [review]
I prefer option 2 as well,
Please pick another color from [this](https://github.com/theforeman/foreman/pull/5137#issuecomment-356959608) list.

**Comment by lizagilman** [comment]
ok, changed - 

image

**Comment by boaz0** [comment]
LGTM :+1: 

test failures not related.
[test foreman]
[test katello]

**Comment by lizagilman** [comment]
[test foreman]

**Review by amirfefer — APPROVED** [review]

**Review by tbrisker — APPROVED** [review]
Tested, works well. Thanks @lizagilman !

**Pull request merged**